### PR TITLE
Merged internal and public PeerConnection classes

### PR DESCRIPTION
**Pull request by piranna**
Here you have an example of my proposal to unify internal and public classes, in this case done with the PeerConnection one. As a result of it, not only the code is almost 100 lines smaller and gets simpler and cleaner and more easier to maintain and expand (adding EventTarget events will be trivial), but also it will waste less memory and there are no private attributes accesibles from outside, making it more secure and standard compliant. I have updated the methods parameters names to the current ones on the specification. For the native PeerConnection events onicecandidate and ondatachannel I needed to remove the function names due to conflicts that provocated an infinite recursion, but this will be fixed when EventTarget events get implemented.

Opinions? Comments? Suggestions? If you agree, I can keep working on adding the EventTarget events on this and apply the same code scheme to the other public classes.

### CODE REVIEW

**Comment by piranna** [comment]
Now PeerConnection class implement W3C EventTarget interface (addEventListener and friends). As I supossed, it was easy :-) GitHub doesn't show the code in a clear way, it's better to check it with Meld, but I can confirm that the tests pass and also that my library is working with this changes (it only left that DataChannels also implement it...).

**Comment by piranna** [comment]
I've also unified the DataChannel classes and added support for EventTarget and it now works with my library, although it give me a segmentation fault (seems DataChannels needs a little bit more love).

**Comment by piranna** [comment]
I've finished to unify all public and private classes according to the spec. Also, this fixes issues #80 and #79 but also I think it fixes issue #8, and will make it easy to fix issues #44 and #37.

File: package.json [hunk 1]
```diff
@@ -22,16 +22,17 @@
   },
   "dependencies": {
     "bindings": "~1.1.1",
-    "node-gyp": "~0.10.10",
+    "eventtarget": "0.0.0",
```
Comment (modeswitch) [inline package.json]: Why not use https://github.com/WebReflection/event-target?

**Comment by piranna** [comment]
I honestly thought about using it, but after studying its code and how to use it I didn't like the style, too much complicated just to don't allow to "inherit from EventTarget object". I find this point just a mental game, no something serious. The thing about mixin mapping by hand the methods using object augmentation instead of proper interfaces (that by the way, it's how W3C APIs are implemented) I find it childish.

The code I'm using instead (https://github.com/shareit-project/eventtarget.js) it's a simple, easy to understand and following the W3C events specification (also working with property events) improved version of another already highly forked one (https://github.com/mrdoob/eventdispatcher.js), and I have been using it for almost two years on my P2P library before moving to Node.js, so it's greatly tested and works. The 0.0.0 version is because until yesterday it was not published on NPM, and since I was doing some tests I didn't put too much attention on the version.

**Comment by piranna** [comment]
Another alternative, obviously, would have been to add the methods directly to the classes...

**Comment by wouldgo** [comment]
Just for a brief recap. Is this PR feasible?
If this will merged I've some (trivial) activities to do (I want to close #56).

**Comment by piranna** [comment]
What things do you want to do? This pull-request can be included directly, but maybe it would need more discussion about if there is discussion this is the way to go (I think so). If you want you can fetch the code from my repo and later merge it if the pull-request gets accepted. Anyway, you can share what activities you want to do related with issue #56 so I can help you.

**Comment by wouldgo** [comment]
Nothing special, atm I see only these:
- lining js-es with the jshint rules that I pushed ten days ago;
- reviewing the tests, you told that they could have race conditions (#67).

**Comment by piranna** [comment]
- Ok, pass JsHint on this pull-request is good to me, do a checkout from my repo and later send me a pull-request, when I merge them changes will be automatically shown here on this pull-request :-)
- I didn't touch the tests, so if you want to work on them, it's better that you work on a checkout from the original code currently on the develop master.

**Comment by modeswitch** [comment]
@wouldgo Feel free to open a separate pull request. This PR can be rebased if necessary.
@piranna I haven't finished reviewing your changes. I'll merge when I have time (probably today or tomorrow).

Since there's more code coming in now, maybe I should add a `stable` branch so that it's easier to merge onto develop. Thoughts?

**Comment by wouldgo** [comment]
Agreed :+1: 

The stable branch imho is a good practice.

**Comment by piranna** [comment]
@modeswitch, don't worry take your time :-) Ask me any doubt you have.

Regarding stable branch: I don't believe too much code is being added, but it's true review time is costly. I would use master branch for this, maybe after milestone 0.1.0 is achieved?

File: lib/datachannelevent.js [hunk 2]
```diff
@@ -1,5 +1,7 @@
 function RTCDataChannelEvent(channel) {
   this.channel = channel;
 }
+RTCDataChannelEvent.prototype.type = 'datachannel';
```
Comment (modeswitch) [inline lib/datachannelevent.js:4]: Why is this necessary?

**Comment by piranna** [comment]
It's for how [W3C events](http://www.w3.org/TR/DOM-Level-3-Events/#interface-Event) works: they are an object and the addEventListener and dispatchEvent methods of the [EventTarget interface](http://www.w3.org/TR/DOM-Level-3-Events/#interface-EventTarget) (implemented in EventTarget.js) use the "type" field to identify what are the listeners that should be dispatched, and on RTCDataChannelEvent this field is equal to "[datachannel](http://www.w3.org/TR/webrtc/#rtcdatachannelevent)".

I have just seen events should fully implement the Event interface instead of being plain objects. Both things are compatible, but it's better to do it when working on issue #8.

**Comment by modeswitch** [comment]
@piranna I'm unable to install `eventtarget`: 

```
npm ERR! Error: shasum check failed for /home/ack/tmp/npm-14413-3rQ7A2t_/1393459358841-0.8358576616737992/tmp.tgz
npm ERR! Expected: ee88e458920ce9217d2c69c2931580a2b67f2962
npm ERR! Actual:   12c422babde122bf240c2afac662b53279468529
npm ERR! From:     https://registry.npmjs.org/eventtarget/-/eventtarget-0.0.0.tgz
npm ERR!     at /usr/lib/node_modules/npm/node_modules/sha/index.js:38:8
npm ERR!     at ReadStream.<anonymous> (/usr/lib/node_modules/npm/node_modules/sha/index.js:85:7)
npm ERR!     at ReadStream.EventEmitter.emit (events.js:117:20)
npm ERR!     at _stream_readable.js:920:16
npm ERR!     at process._tickCallback (node.js:415:13)
```

**Comment by piranna** [comment]
Strange error, don't understand why the sha doesn't match. I'll try to see later what's happening, but I'll be busy the next days at work, so maybe not before weekend... :-/

**Comment by piranna** [comment]
I've tried this myself and got the same error, but while opening an issue on NPM, it worked. It seems it was a temporal malfailure of NPM platform, please try again. If it doesn't work, uninstall eventtarget previously to install it again to be sure and keep me informed of your results.

**Comment by modeswitch** [comment]
@piranna I'm still getting that error.

**Comment by modeswitch** [comment]
@piranna Can you try pushing an updated version to NPM?

**Comment by piranna** [comment]
It was working for me, I've removed the package from ~/.npm cache and now it fails. I've uploaded a new dumb version and it was still failling, but after changing the version on package.json of node-webrtc from "0.0.0" to "~0.0.0" it was able to download it, please try again after changing this line.

**Comment by modeswitch** [comment]
OK, works now. Can you update your pull request to rely on eventtarget ~0.0.1?

**Comment by modeswitch** [comment]
Also, I'm getting test failures now: Error: CreateAnswer can't be called before SetRemoteDescription.

**Comment by piranna** [comment]
Pull-request updated. Can you send me the test trace? I have them passing here... Maybe you have updated the libWebrtc code?

**Comment by modeswitch** [comment]
```
> wrtc@0.0.9 test /home/ack/work/node-webrtc
> node test/all.js

TAP version 13
# create a peer connection
-> TRACE: Call::static v8::Handle<v8::Value> PeerConnection::New(const v8::Arguments&)
<- Call::static v8::Handle<v8::Value> PeerConnection::New(const v8::Arguments&)
ok 1 created
# createOffer function implemented
ok 2 implemented
# can call createOffer
-> TRACE: Call::static v8::Handle<v8::Value> PeerConnection::CreateOffer(const v8::Arguments&)
-> TRACE: Call::virtual void CreateOfferObserver::OnFailure(const string&)
-> TRACE: Call::void PeerConnection::QueueEvent(PeerConnection::AsyncEventType, void*)
<- Call::static v8::Handle<v8::Value> PeerConnection::CreateOffer(const v8::Arguments&)
-> TRACE: Call::static v8::Handle<v8::Value> PeerConnection::CreateOffer(const v8::Arguments&)
<- Call::void PeerConnection::QueueEvent(PeerConnection::AsyncEventType, void*)
<- Call::virtual void CreateOfferObserver::OnFailure(const string&)
-> TRACE: Call::virtual void CreateOfferObserver::OnFailure(const string&)
-> TRACE: Call::void PeerConnection::QueueEvent(PeerConnection::AsyncEventType, void*)
<- Call::void PeerConnection::QueueEvent(PeerConnection::AsyncEventType, void*)
<- Call::virtual void CreateOfferObserver::OnFailure(const string&)
<- Call::static v8::Handle<v8::Value> PeerConnection::CreateOffer(const v8::Arguments&)
-> TRACE: Call::static void PeerConnection::Run(uv_async_t*, int)
   TRACE: evt.type : 2
not ok 3 Error: CreateAnswer can't be called before SetRemoteDescription.
  ---
    operator: error
    expected: 
    actual:   {}
    at: PeerConnection.onerror (/home/ack/work/node-webrtc/lib/peerconnection.js:80:24)
    stack:
      Error: CreateAnswer can't be called before SetRemoteDescription.
  ...
   TRACE: evt.type : 2
<- Call::static void PeerConnection::Run(uv_async_t*, int)
```

**Comment by piranna** [comment]
Seems it's some change done on libwebrtc, all it's fine here. I'm executing `npm install` to update it and see if it fails, but by the way, I think the error makes sense, since createAnswer() can't be able to work properly until it have the content of the remote offer SDP, so setRemoteDescription() should be called first...

**Comment by modeswitch** [comment]
Oh, this is also failing on develop, so not your fault. Maybe a broken test. I don't have time to debug now, maybe @DamonOehlman does?

**Comment by piranna** [comment]
Lol :-P Then confirmed it's a libWebrtc issue :-)

**Comment by piranna** [comment]
```
1..49
# tests 49
# pass  49

# ok
```

After executing `npm install` and update the libWebrtc code, the test are passing here... :-/

**Comment by modeswitch** [comment]
Huh, I rebuilt and now it's working for me too. Hurray!?

**Comment by piranna** [comment]
Sort of :-P I bet they uploaded to the repository broken code (again...)

**Comment by modeswitch** [comment]
Well, they SCTP changes have landed in the stable branch yet. Once that happens I'll point this project at the stable branch and the upstream code won't change as often.
